### ISSUE TITLE
Problem with imported cylindrical models

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#505]
While python is able to import some cylindrical models, it is having problems with with imported cylindrical model that are intended to have very few cells int he wind. When you look at the outputs (created with windsave) the cells appear be in the wind, but cells that should be in the wind have no photons.

**Comment by kslong** [comment]
@ssim   At least part of the problem was that Python was not properly establishing the boundaries of the wind with a cylindrical imported model, once this was fixed I was able to run a version of the files that you sent, but only  after I  changed some of the velocities so they were not all of the same.  When they were the same, Python seemed to get stuck, I think because of something in the velocity gradient tensor.   

So that we can regenerate this example, here are all of the files.  

The original file which Stuart created.

[Ring-test.orig.txt](https://github.com/agnwinds/python/files/3091831/Ring-test.orig.txt)

.. as modified

[Ring-test.txt](https://github.com/agnwinds/python/files/3091834/Ring-test.txt)

and the pf file

[ring_import.pf.txt](https://github.com/agnwinds/python/files/3091836/ring_import.pf.txt)

Could you check that this works for you (by pulling the latest dev version)

Before this issue is closed, ksl needs to check that we do not have similar problems importing other types of grids.

**Comment by kslong** [comment]
There is definitely the same problem for imported rtheta grids.  The boundaries for spherical models may be OK.

**Comment by kslong** [comment]
I ran the original version (with velocities all the same) of the imported model overnight, and it got stuck almost immediately, once the photons began to be translated though the wind.  We definitely should locate this error,  It may be related to #385 which I recently closed.

**Comment by ssim** [comment]
Thanks, Knox - I can confirm that this is all as it is for me now too - with the latest "dev" version I can also now get it to work, but also have the same issue you mention (namely that it gets stuck immediately if I set the velocities to all being the same - i.e. I think the way I was trying to set it up was corresponding to constant circular rotation speed).

It's also now the case that I only need to set one of the zones to be "inwind = 0" to get the desired behaviour - I no longer need to set a little 2x2 cube to be inwind - so that is also good.

**Comment by kslong** [comment]
This I believe is where photons are being trapped when the velocity gradients are zero.

![image](https://user-images.githubusercontent.com/2530901/56380105-e25aae00-61d6-11e9-877a-c885c257866e.png)


Note that for an example that I trapped nnscat was 774e6, so definitely we need to do something in this case.

**Comment by kslong** [comment]
@jhmatthews @ssim  we have had some email discussions about this problem and this comment is to bring this conversation back into github so we can resolve it in the end.  

The problem we currently have is in the "normal" mode using anisotropic thermal scattering.  It arises in the loop above in situations where dv/ds is negligibly small, so tau is effectively infinite.  These leaves z effectively zero and we never escape the loop once it is entered.

The problem has come up before.  As James notes, there are associated errors where we tried to deal with this from around 2014  #82 and later #342 #365   Most of the work in these codes was intended to deal with issues associated with macro atoms although it was noted that it also affected non-macro atoms.

The code comments include something having to do with a safety net, but Knox was not able to discover what this was.  (I had trouble getting the .c files from this time period into my git repository). Perhaps James can say.

In principle, the immediate problem could be avoided if we simply limited the maximum number of times we went go through the do loop ... but nnscat is used to diminish the photon weight in the non-macro atom case.  (This was part of the point of this mode)

(If Knox remembers correctly, at one point we estimated the number of scatters from tau_max and an assumption of how tau varied with angle with respect to this, and we generated the direction directly.  We changed this because it was pointed out that here is no reason to believe the velocity tensor is simple). 

It is not entirely obvious what one should do about all of this, especially for wind models where this comes up a lot.   For a simple velocity field, our model has just as many photons going in the "forward" as "reverse" direction, where as I suspect that in reality when one encounters a Sobolev surface, most photons would come out in the reverse/reflected direction.  I also expect that frequency shifts from scatter to scatter would reduce the number of times a photon would scatter.

That said, Knox's base requirement is that this does not cause the program to fall over, so what I propose to do is to set a maximum number of times for the photon to go through the loop (10,000) and choose a direction based on the time we got closest to escaping.  I'll also try to capture the number of times we reach this limit.  

Then I propose to close this issue, which is misnamed for this problem and open a new one to address the larger issues it raises.  But I would appreciate everyone's comments.

**Comment by kslong** [comment]
I have made modifications on the dev branch, which drop out of the loop discussed above when one traverses the while loop more than 10,000 times.  In this instance, an error is recorded and the photon direction is set to the direction where the lowest tau was encountered.   

This fixes the hang problem, and the error summary prints out the number of times we have encountered this error, which for Stuart's example turns out to be about 2700 times.  

Therefore I am closing this issue, and starting a new one #507 to specifically address the physical question of what we ought to be doing.
